nexus/src/app/background/services_propagation.rs

```diff
@@ -0,0 +1,290 @@
+            // to do that while we go off (potentially for a while) attempting
+            // to update the servers.
+            let service_config = self.rx_config.borrow().clone();
+
+            let service_config = match service_config {
+                Some(s) => s,
+                None => {
+                    warn!(
+                        &opctx.log,
+                        "Service propagation: skipped";
+                        "reason" => "no known services"
+                    );
+                    return json!({ "error": "no known services" });
+                }
+            };
+
+            // Propagate the config to all of the sled agent servers.
+            let log = &opctx.log;
+            let result = services_propagate(
+                &log,
+                service_config,
+                self.max_concurrent_server_updates,
+            )
+            .await;
+
+            // Report results.
+            if result.iter().any(|r| r.is_err()) {
+                warn!(&log, "Sled agent service propagation: failed");
+            } else {
+                info!(&log, "Sled agent service propagation: done");
+            }
+
+            serde_json::to_value(
+                result
+                    .into_iter()
+                    .map(|r| r.map_err(|e| format!("{:#}", e)))
+                    .collect::<Vec<_>>(),
+            )
+            .unwrap_or_else(|error| {
+                json!({
+                    "error":
+                        format!("failed to serialize final value: {:#}", error)
+                })
+            })
+        }
+        .boxed()
+    }
+}
+
+async fn services_propagate(
+    log: &slog::Logger,
+    service_config: SledServices,
+    max_concurrent_server_updates: usize,
+) -> Vec<anyhow::Result<()>> {
+    stream::iter(service_config.into_iter())
+        .map(|(sled_id, services)| async move {
+            services_propagate_one_sled(log, sled_id, services).await
+        })
+        .buffered(max_concurrent_server_updates)
+        .collect()
+        .await
+}
+
+async fn services_propagate_one_sled(
+    log: &slog::Logger,
+    sled_addr: SocketAddr,
+    services: ServiceEnsureBody,
+) -> anyhow::Result<()> {
+    let url = format!("http://{}", sled_addr);
+    let log = log.new(o!("sled_server_url" => url.clone()));
+    let client = sled_agent_client::Client::new(&url, log.clone());
+
+    let result = client.services_put(&services).await.with_context(|| {
+        format!("failed to propagate services to sled {}", sled_addr,)
+    });
+
+    match result {
+        Err(error) => {
+            warn!(log, "{:#}", error);
+            Err(error)
+        }
+        Ok(_) => {
+            info!(
+                log,
+                "Sled Services Propagated";
+                "address" => sled_addr,
+                "generation" => services.generation.to_string(),
+            );
+            Ok(())
+        }
+    }
+}
```

Comment:
Something to be aware of: One bad sled could block us here? Maybe we should be tracking this "task" durably across the task "activate" calls, and we could drop if it we no longer care about this sled.

Reply:
maybe punt on this, DNS propagation has the same problem